### ISSUE TITLE
Use of go.mod fails with golang 1.13+ unless run as root

### ISSUE THREAD

**Issue by electrofelix** [asottile/setuptools-golang#49]
It appears newer versions of golang (not entirely sure when this was introduced) remove write permissions for the cache populated, which because it is under a temp directory context will trigger an error on attempt to remove.

```
$ CGO_CFLAGS='-I<project_dir>/.tox/py37/include -I<homedir>/.pyenv/versions/3.7.7/include/python3.7m' CGO_LDFLAGS=-Wl,--unresolved-symbols=ignore-all GOPATH=<usertmpbasedir>/tmpqb7gjsft go build -buildmode=c-shared -o <project_dir>/build/lib.linux-x86_64-3.7/<my-library>.cpython-37m-x86_64-linux-gnu.so
error: [Errno 13] Permission denied: '<usertmpbasedir>/tmpqb7gjsft/pkg/mod/google.golang.org/protobuf@v1.23.0/compiler/protogen/protogen_test.go'
```
Sanitizing the output because the project is an internal work based one

```
ls -la <usertmpbasedir>/tmpqb7gjsft/pkg/mod/google.golang.org/protobuf@v1.23.0/compiler/protogen/
total 68
dr-xr-xr-x 2 electrofelix users  4096 Sep 21 14:49 .
dr-xr-xr-x 3 electrofelix users  4096 Sep 21 14:49 ..
-r--r--r-- 1 electrofelix users 45926 Sep 21 14:49 protogen.go
--w------- 1 electrofelix users 10685 Sep 21 14:49 protogen_test.go
```

Notice no write access on the directory, even run manually:
```
rm -f <usertmpbasedir>/tmpqb7gjsft/pkg/mod/google.golang.org/protobuf@v1.23.0/compiler/protogen/protogen_test.go
```
gets the same error of `Permission denied`

Though perhaps given that downloading the go dependencies pristine each time it might be more performant to allow the user to reuse their existing golang cache or even only download to the temp directory those files that are not already available under GOPATH as it can be a list of paths instead of a single path.

**Comment by asottile** [comment]
can probably use the same trick [that pre-commit uses](https://github.com/pre-commit/pre-commit/blob/1b3d14237a87abfba41d7d1f890ea703153b15d8/pre_commit/util.py#L251-L268) if you'd like to send a patch and bump the tested go versions to the latest two (those are all that upstream supports)

**Comment by electrofelix** [comment]
I've updated the title, it appears to be an issue where use of go.mod or a module cache is involved where the directories were made not writable at least as far back as Aug 2018 https://github.com/golang/go/issues/27161#issuecomment-433102849. It appears to be only use of go.mod that triggers this behaviour for this module, possibly because that is the only one that results in a module cache being populated under a temporary directory. This should mean that the current versions of go being tested are just fine, so I added a test that triggered the issue locally for me, and I believe should trigger it (if you remove the fix) on go from 1.11+
